src/app.js

```diff
@@ -0,0 +1,369 @@
+        display: true,
+        position: 'right',
+        id: 'cost-axis',
+        scaleLabel: {
+          display: true,
+          labelString: 'Invocation Cost',
+        },
+        // grid line settings
+        gridLines: {
+          drawOnChartArea: false, // only want the grid lines for one axis to show up
+        },
+      },
+    ],
+  },
+};
+
+export class App {
+  constructor(store) {
+    this.store = store;
+    this.canvas = document.querySelector('canvas');
+    this.reportElem = document.getElementById('report');
+    this.errorElem = document.getElementById('error');
+
+    // App State
+    this.data1 = null;
+    this.data1URL = '';
+    this.data2 = null;
+    this.data2URL = '';
+    this.legend1 = '';
+    this.legend2 = '';
+
+    this.store.addListener(this.show.bind(this), this.error.bind(this));
+  }
+
+  setData1(parts) {
+    this.data1 = parseHash(parts);
+    this.data1URL = getURLForHash(parts.join(HASH_DELIM));
+  }
+
+  setData2(parts) {
+    this.data2 = parseHash(parts);
+    this.data2URL = getURLForHash(parts.join(HASH_DELIM));
+  }
+
+  error(err) {
+    this.reportElem.style.display = 'none';
+    this.errorElem.style.display = 'flex';
+
+    switch (err) {
+      case 'empty':
+        // eslint-disable-next-line no-case-declarations
+        const sampleURL = getURLForHash(
+          encode([128, 256, 512, 1024], Int16Array) +
+            HASH_DELIM +
+            encode([36555.39, 18441.58, 8539.34, 3870.76]) +
+            HASH_DELIM +
+            encode([0.00007612, 0.00007696, 0.00007155, 0.00006489]),
+        );
+        document.getElementById('error-title').innerHTML = 'Error: URL must contain data in the hash!';
+        document.getElementById(
+          'error-msg',
+        ).innerHTML = `Your URL is ${window.location} and should be something like <a href="${sampleURL}" class="dont-break-out">${sampleURL}</a><br>Please refer to the <a href="${DOC_URL}">documentation</a>.`;
+        break;
+
+      case 'malformed':
+        document.getElementById('error-title').innerHTML = 'Error: malformed URL parameters!';
+        document.getElementById(
+          'error-msg',
+        ).innerHTML = `Please check your URL parameters, refer to the <a href='${DOC_URL}#url-format'>documentation</a>.`;
+        break;
+
+      default:
+        // general error
+        document.getElementById('error-title').innerHTML = 'Error';
+        document.getElementById('error-msg').innerHTML = 'Please check your URL parameters';
+    }
+  }
+
+  show(hash) {
+    const hashParts = hash.split(HASH_DELIM);
+    const hashPartsLen = hashParts.length;
+    this.legend1 = '';
+    this.legend2 = '';
+
+    // There must be a single set of values (3 parts), OR, the comparison-set-of-values (8)
+    if (hashPartsLen !== NORMAL_HASH_LEN && hashPartsLen !== COMPARISON_HASH_LEN) {
+      this.error('malformed');
+    }
+
+    try {
+      this.setData1(hashParts.slice(0, 3));
```

Comment:
We're doing this even if `hashPartsLen === COMPARISON_HASH_LEN`, right? Should we check that `hashPartsLen === NORMAL_HASH_LEN` to avoid setting data twice?

Reply:
There are two different "data" properties on the App class: `data1` and `data2`. We always try to set `data1`, and if it fails, raise an error. On line 143, we check `if (hashPartsLen !== NORMAL_HASH_LEN && hashPartsLen !== COMPARISON_HASH_LEN)` then raise an error. So by the time we reach L148, we know that we either have `NORMAL_HAHS_LEN` or `COMPARISON_HASH_LEN`.

We don't set `data1` twice. We set `data1` and `data2` once, each.

`data1` and `data2`  are not the best names, but they are passable IMHO.